### ISSUE TITLE
invalid expressions in const-expr and override-expr should be error (no later than pipeline creation time)

### ISSUE THREAD

**Issue by dneto0** [gpuweb/gpuweb#3253]
Example:

```
const v2 = vec2(0,1);
const badindex = v2[3]; // out of bounds
const badint = 1/0;   // integer divide by zero
const badfloat = 1.0/0.0; // would generate infinite AbstractFloat
const badsqrt = sqrt(-1.0);  // bad value. NaN should be disallowed.

```

In the past we discussed the out-of-bounds case and decided to make *runtime* evaluation consistent with evaluation at creation-time error.

But now we have the spectre of evaluating `static_assert`, and that *should* be as consistent as possible between implementations.  This has changed Google's stance and we now would like to have const-expr evaluation raise errors *before runtime*, and should not be downgraded to a "dynamic error".

Whatever choice is made here, the rule for `override` expressions should be the same as the rule for `const` expressions.
As an example, the Tint's initial implementation plan is to add a pass to lower override expressions to const-expressions at pipeline creation time. So we really would like 'override' be treated as 'const'

**Comment by dneto0** [comment]
Note that this proposal would lead to some inconsistency between const-expr evaluation and runtime-expression evaluation.

E.g.  
```
fn foo() { 
   var one: i32 = 1;
   let runtime_div_by_zero  = one / 0;  // we've polyfilled this with some defined meaning (can't remember)
   const div_by_zero = 1/0;  // Proposing this be shader-creation error.  
 }
```
But we think the tradeoff is that these are "bad" expressions and should generate errors when we possibly can evaluate them in the compiler (before hitting the GPU).

**Comment by litherum** [comment]
Our position on this:
- If the invalid constexpr expression involves an out-of-bounds access, then it should be a compile-time error. The author is bad and they should feel bad.
- Otherwise (for cases like division by zero, or `arcsin(2)`), they should be valid and just obey the behavior that WGSL defines. That's just how WGSL works.

**Comment by litherum** [comment]
> If the invalid constexpr expression involves an out-of-bounds access, then it should be a compile-time error. The author is bad and they should feel bad.

This is actually super powerful, because authors don't have to opt-in to constexpr. For the most part, any operations they do that only involves literals will get these compile errors if they mess it up. 👌

**Comment by litherum** [comment]
> `arcsin(2)`

Note the spec right now doesn't define what this returns.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-08-02-minutes#invalid-expressions-in-const-expr-and-override-expr-should-be-error-no-later-than-pipeline-creation-time-3253">WGSL meeting minutes 2022-08-02</a></summary>

* JB: Seems consensus on making the out-of-bounds case should be such an error.
* JB: When you have an inconsistency and one of the alternatives is “hard stop”, then that’s less of a portability problem because programmer sees it.
* JB: If there is no good usage of a feature then it’s ok to ban it. If a programmer can reasonably rely on a specified behaviour then it’s more problematic. E.g. JS  obj.x where the object does not have ‘x’ property.  JS programmers rely on that.  So we should look at it case by case. 
* MM: With you about splitting into two cases: out of bounds, vs math operations.  Agree out of bounds accesses are definitely wrong.  There are so many wildly different results that can occur, which are legal by our language.  So the philosophy of “are people doing this on purpose” is not the right test. Nobody writes “1/0” intentionally. It’s sensible to honor the “x/y” semantics from the main language.  We think the right test is “is the set of potential behaviours of this operation wildly large, or is it fairly narrowly scoped”.  And arcsin(2) (is expected to) have narrow behaviour.
* DG: Another test we also considered: If the program would stop, then at compile time, that should be an error.
* MM: Right now, we don’t let the program stop, but I have filed an issue to add that because I thought that was allowed.
* MM: I filed this issue: [https://github.com/gpuweb/gpuweb/issues/3272](https://github.com/gpuweb/gpuweb/issues/3272)
* KG: I do write 1/0 sometimes.
* AB: Think that line is arbitrary.  No other language defines “x/0” .We decided to do that “just because” so that’s not enough impetus on its own.  The mathematical domain questions are still more choices by us, if we make them.  The OOB access on buffers is from previous behaviour of APIs.  The cutline feels arbitrary from a previous arbitrary choice.
* MM: Depends what you mean by an “error”.  E.g. in WGSL indexing off the end of the array _is _ an error.  But in JS it’s not an error.  
* AB: JS decides to give it the behaviour.  Depends one which array it is, determines whether we give the array access the meaning of an error.  I’ve seen HLSL shaders in the wild relying on OOB indexing being 0-valued. 
* MM: Not special to constexpr.
* AB: We try to avoid undef behaviour in underlying APIs. That’s the driving reason, not the portability reason.
* MM: Doesn’t matter why, but folks are going to rely on it.
* AB:So we can choose which way to set it. Incentivize away from bad math being a useful result. That’s more useful that defining an arbitrary value.
* MM: Why discourage them from bad math, if it’s defined.
* JB: If we make it a compile time error, we can walk it back in the future.
* JB: Yes, we can give these things values, and folks might find ways to use them. But we should discourage them from using it because it’s bad code. The code will rely on a corner of the spec. It’s mean to code readers.
* TR: If there are cases where a tool is collapsing code, and ends up that because it’s able to consteval something that it wouldn’t before, then it will fail to compile, and fail to run the same thing that potentially could run with defined behaviour.  And indexing out of bounds is used in practice because of the defined rules.
* MM: This is a real issue. A code generator could hit on this case; e.g. do substitution.
* DN:  Agree with Jim’s argument.  But additionally array indexing is a “computer” concept, but trig math is hundreds of years old, so we should embed in that cultural practice. Which means to me arcsin(2) is nonsensical and the lang should not define it.  As for consistency of consteval vs. runtie evaluation, that’s an argument for reopening some of the choices for math functions at runtime too. 
* **KG: Let’s come back next week.**

</details>

**Comment by jimblandy** [comment]
> - If the invalid constexpr expression involves an out-of-bounds access, then it should be a compile-time error. The author is bad and they should feel bad.
> - Otherwise (for cases like division by zero, or arcsin(2)), they should be valid and just obey the behavior that WGSL defines. Those operations aren't errors; they're just how WGSL works.

I sort of like what you're aiming for here, but I think "they're just how WGSL works" doesn't carry the load you're hoping it will, here. WGSL says how out-of-bounds accesses work, too.

My preferred criterion is "do people write code that uses this behavior on purpose". This has the advantage of drawing the line between intentional behavior and unintentional mistakes. Things we expect to be usually unintentional mistakes should be static errors when they occur in constant expressions.

My preferred criterion has the disadvantage of being dependent on unavailable knowledge of 'what people do', and even when we somehow obtain that knowledge anyway, my criterion is subjective ("Fine, some mutant does actually write code that uses that, but respectable persons would never").

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-08-09-minutes#invalid-expressions-in-const-expr-and-override-expr-should-be-error-no-later-than-pipeline-creation-time-3253">WGSL 2022-08-09 Minutes</a></summary>

* JB: Substituted one problematic criteria for another one. We defined how out of bounds works, so we can’t say this is how wgsl works to define the cases. Need to make judgment call on what would be likely used on purpose and what by accident in reasonable code.
* MM: After talking about trapping, our logic came from if you immagine trapping being added then one consequence of out of bounds is program stopping. Thought that was strong enough to have implications on this issue. But for something like arcsin(2) there is no way… coming from idea we accept trap
* JB: Should we not trap on arcsin(2). If we have trap we should trap on that, it’s meaningless, it’s a mistake.
* MM: Tough pill to swallow. Can’t articulate why.
* JB: LIke idea of aligning trap and static errors in constexpr.
* MM: One thing important with trap, no trap is ever forced, no situation where compiler has to trap. With arcsin(2) would compiler be forced to trap?
* JB: Assume similar language for domain errors as out of bounds where you can choose 1 of multiple behaviours. Think this is ok, if you stop execution and tell the user about it. Not the same category as giving various answers
* AB: How do you tell the user? Sounds like pipeline keeps working and gives, something. Feels worse from portability argument
* KG: Have option as a UA but not required. Behave like you did something weird in GLSL and get blackout bliz
* KG: One thing for our heuristic, if it’s easy to choose, two behaviours user might want and we don’t know which they do, if we pick our default so user becomes aware and can fix it. For out of bounds access at compile time, get error and can tell them to add clamp. But they don’t get wrong behaviour at compile time. Strictly can’t share the code and can’t expect inlining.
* DN: If can summarize, then bailing out loudly is a good option as then the user can choose how to react. Want to +1 what JB wrote in issue which is that “using the behaviour on purpose”. To me arcsin(2) isn’t sensible and wouldn’t be used on purpose. Those aren’t in the domain of those math functions and the computer should approximate the math. For a runtime vs compile time inconsistency for me one guide is if you get puzzled use a constant value and the compiler will tell you if it’s bad. For the math domain things, hopefully heading to world with gpu having nan and full support. CPU will have nans and we should find those cases. In future date we can cut off old behaviour and make it more strict. Think we should head towards world where we enforce what IEEE gives us.
* AB: So, sounds like we would have compile time error if make NaN in const expr. IEEE calls that a result. Can buy where one domain produces another but don’t think you can get to a place where it isn't’ an error without it being weird. Either don’t use NaNs or have full nan support. We don’t now because we wanted to be consistent with abstract float.
* KG: Like idea of requiring compile time NaN is error. Think that’s a cool way to get most of the benefit.
* AB: THink it’s OK but calling it an IEEE thing isn’t right as it isn't a problem in IEEE.
* KG: Think it’s ok, we don’t have IEEE floats, at compile time we still don't, just a different flavour, which is NaN is compile time error.
* MM: So eventually, will add IEEE 754 compat extension and that extension would relax this compile error?
* AB: Unlikely we get to full IEEE
* MM: It’s an extension
* AB: Don’t think there is a lot of hardware to get full, caveats on it.
* KG: Consider a more float option
* MM: Could go way way further and still get a bunch of hardware.
* KG: Semantics if we call it IEEE.
* KG: In favour of strawperson which does compile time NaN generation as error
* DN: Think already on board with out of bounds is an error
* KG: Want to protect 1/0 ability
* MM: Little confused. Coming from expectation that in WGSL will define arcsin(2) is something that isn’t NaN
* DN: Don’t like that but can litigate later
* MM: Because of the policy just describe, if a compilation could make a NaN that’s a compile error. If we define arcsin(2) to not be nan it’s portabile and well define it wouldn’t be hit that conditional and be a compile error
* KG: Might want to use that at compile time
* AB: You’ve picked some special rules to follow, why out of bounds and not arc sin
* MM: Philosophy, if many possibilities of behaviour, lots for out of bounds, because so many options that’s the boundary condition.
* KG: But if always one solution?
* MM: That’s just how the language works. If one possibility is the program stops running, then pretty strong hint something is wrong.
* DS: So we define arcsin as doing anything?
* AB: We sort of do
* MM: Do it like division?
* AB: Some hardware loads away NaN so it won’t exist.
* KG: Not proposal. 2 things, compiletime and runtime. Runtime has no NaNs, compile time has NaNs and if you end up assigning that NaN out to a constexpr result that’s an error.
* AB: That’s fine, but not what MM is saying, MM says give answer like 1 or 0
* MM: arcsin is already really slow, so checking if > pi or &lt; -pi doesn’t make slower so it’s a win
* DN: Reduces readability as it’s a trick in the spec.
* JB: Not something folks should do on purpose
* KG: But have to be aware in practise
* MM: What is it
* JB: Depending on arcsin on out of domain inputs, Folks should not do that
* AB: What is cut off between slow and fast. Is tan slow or fast? Any trig?
* MM: One criteria, is does HLSL have an intrinsic for it?
* DN: atan2 vs atan, that’s a sharper edge
* KG: Some time just have to choose and swallow doubt and pick something.
* MM: Could also say, does it take longer then 100 cycles.
* KG: Even harder to litigate this then just decide atan2 is slow and tan isn’t. We should think about this more.
* AB: Should also consider portability is being raised but trap is making things less portable. Want to know where we pick and choose
* JB: Withdraw what I said, as assuming it was a hard stop, if program continues to run then it isn't’ a portability problem.
* MM: Can defend portability, there is already a bunch of options about what might happen with out of bounds and this is an anti-pattern in wgsl. Taking something that in other languages is worse behaviour, in others it’s undefined, and we’re talking the set of possibilities and making them smaller. Have a list of 7-8 behaviours for out of bounds, and proposing making that 8-9 which is better then other languages.
* AB: Except for spooky action at a distance where I’ve done an discard and do this as well. Extra side effects, not one access.
* KG: **We’ll think about it.**

</details>

**Comment by jimblandy** [comment]
Here's one approach we might take for the math functions whose domain is not all of `f32`, that provides a hook on which to hang both NaN handling and early error reporting:

- For each math function, specify its domain. The domain of `asin` (pace Wolfram Alpha) would be [-1, 1].
- In common introductory text for math functions, say:
    - Passing a function a value outside its domain results in a "domain error".
    - In const expressions, domain errors are shader-creation errors.
    - In pipeline expressions, domain errors are pipeline-creation errors.
    - In runtime-evaluated expressions:
        - On platforms where NaNs work, functions should return a NaN if a domain error occurs.
        - On other platforms, they should return some value chosen by the implementation at run time.

Introducing the "domain error" terminology lets us handle all the math functions in a consistent way.

I feel that WGSL should not try to make math functions behave usefully in the case of domain errors. I don't want to encourage people to write code that depends on their behavior.

Requiring "some value chosen by the implementation at run time" does not preclude the implementation choosing it at some earlier phase, or even always returning the same value. It does preclude exhibiting undefined behavior: poison- or undef-like values are not permitted. If the platform facilities would exhibit undefined behavior, WGSL implementations would need to polyfill them to avoid it. This is a feature.

**Comment by jimblandy** [comment]
Since WGSL floats do include infinities and positive and negative zero, I would argue that the domain of floating-point division should include positive and negative zero as divisors, since the output is well-defined and meaningful. In other words, `1/0` should not be a domain error, unlike `asin(2)`.

**Comment by jimblandy** [comment]
Since WGSL floats do include infinities and positive and negative zero, I would argue that the domain of floating-point division should include positive and negative zero as divisors, since the output is well-defined and meaningful. In other words, unlike `sqrt(-1)`, `1/0` should not be a domain error.

**Comment by dneto0** [comment]
I very much agree with @jimblandy 's "domain error" approach.

This is very well aligned with existing practice in the C/C++ standard library.  Eg. many descriptions for math functions on cppreference.com describe domain error cases.  C/C++ implementations need not implement all of IEEE 754, and IEEE 754 can be more strict.

**We should not break mathematics**.

It would be bad for people to end up relying on sqrt(-1) having some particular value that we chose because we wanted to avoid undefined behaviour at _runtime_.  As Jim put it earlier, it makes the code hard to read.

I did a first pass to list the domains of the numeric functions in WGSL.  I consulted various references.
I counted an input as being in a valid domain if it does not provoke the output to be NaN.



Function | Domain (when is the result not NaN)  (total is [-inf,inf]) |   |  
-- | -- | -- | --
abs | total |   |  
acos | [-1,1] |   |  
acosh | [1,inf) |   |  
asin | [-1,1] |   |  
asinh | total |   |  
atan | total |   |  
atanh | (-1,1) | poles at -1, +1 | FE_DIVBYZERO
atan2 | domain is total, except x=0, y=0See https://en.cppreference.com/w/cpp/numeric/math/atan2 C library: If IEEE math supported, domain error does not occur for {x=0 and y = 0};   NaN returned if either X or Y are NaN |   |  
ceil | total |   |  
clamp | total |   |  
cos | (-inf,inf) |   |  
cosh | total |   |  
countLeadingZeroes | total |   |  
countOneBits | total |   |  
countTrailingZeros | total |   |  
cross | derived (total) |   |  
degrees | total |   |  
determinant | derived (total) |   |  
distance | derived from length |   |  
dot | total |   |  
exp | total |   |  
exp2 | total |   |  
extractBits signed | total |   |  
extractBits unsigned | total |   |  
faceForward | total |   |  
firstLeadingBit signed | total |   |  
firstLeadingBit unsigned | total |   |  
firstTrailingBit | total |   |  
floor | total |   |  
fma(x,y,z) | everywhere, except:x=0, y=infx=inf, y=0x*y=inf and z = -x*yx or y are NaNz NaN, x*y are not 0*Inf or Inf*0See https://en.cppreference.com/w/cpp/numeric/math/fma |   |  
fract | total |   |  
frexp | total |   |  
insertBits | total |   |  
inverseSqrt | (0,+inf)  what about infinity? |   |  
ldexp | total | +/- infinity returns itself |  
length | total |   |  
log | x >= 0 |   |  
log2 | x >=0 |   |  
max | total |   |  
min | total |   |  
mix | total |   |  
modf | total | like C++ std::modf(x,1.0) which has domain error only when y=0; for fmodf(x,y) |  
normalize | ???  vector not equal 0? | Can blow up for very small vectors |  
pow(base,exp) | total except:base finite and negative, exp finite and non-integerbase zero and exp zerobase zero and exp negative: domain or pole errorIEEE has tons more cases, some may divide by zeroSee https://en.cppreference.com/w/cpp/numeric/math/pow |   |  
quantizeToF16 | total |   |  
radians | total |   |  
reflect | total (inherited) |   |  
refract | total (inherited) |   |  
reverseBits | total |   |  
round | total |   |  
saturate | total |   |  
sign | total |   |  
sin | (-inf,inf) |   |  
sinh | total |   |  
smoothstep | total |   |  
sqrt | (-0, inf] |   |  
step | total |   |  
tan | (-inf,inf) |   |  
tanh | total |   |  
transpose | total |   |  
trunc | total |   |  
pack* | total |   |  
unpack* | total |   |

**Comment by dneto0** [comment]
In the above, I did not distinguish between +0.0 and -0.0

**Comment by litherum** [comment]
First of all, it's good that so many functions are non-NaN over ~all of their domain.

It looks like these operations produce NaN for roughly half of their domain, or more:

- `acos`
- `acosh`
- `asin`
- `atanh`
- `inverseSqrt`
- `log`
- `log2`
- `sqrt`

2 responses:
1. It's not really okay to have so much undefined behavior here at runtime. Both JavaScript and Web Assembly go way farther than allowing any implementation-defined value to be produced at runtime from these operations.
2. From our perspective, the runtime (implementation-defined) behavior is way more important to fix than the compile-time behavior.

We'd like to split this issue into two halves:

1. What should the runtime-behavior for these operations be, and
2. What should the compile-time behavior be.

We believe the answer to the second question depends on the answer to the first.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-08-16-minutes#invalid-expressions-in-const-expr-and-override-expr-should-be-error-no-later-than-pipeline-creation-time-3253">WGSL 2022-08-16 Minutes</a></summary>

* MM: In discussions, runtime is more important then compile time behaviour. Think the compile time behavour is dependant on runtime. Compile time behaviour depends on runtime behaviour. Want to split issue and discuss runtime first.
* KG: What about the runtime behaviour?
* MM: Not ok to have so many operations that are undefined for roughly half or more of the domains. Made a list of functions that fit that criteria. Would like to make them more defined. Would pick some value for them to have a total domain.
* KG: Seems reasonable.
* AB: Get into all sorts of issues like how expensive is it to support this behaviour
* KG: Key should be focusing on the in bounds should be what is expected. Out of bounds doesn't’ matter so much as long as consistent.
* AB: Matters if you have to treat it the same way. If doing test of select operation all the time using builtin. Expect sqrt to be fast, but is much slower as we’re testing the domain.
* MM: We can get into details of that, think case is clear for inverse trig. Going to be slower everywhere so adding clamp or select wont’ be observable. Those are obvious win, can discuss others.
* KG: Yup. Think we’d want proposals for what to do.
* MM: Strawperson proposal is implementation fo sqrt is actually sqrt(max(x, 0))
* KG: That or abs.
* MM: Right, would be sameish cost, would be fine also
* AB: If x is close to zero, such it rounds, you’d have to take 0 out of the domain. Something just around zero do you round the x?
* KG: SHouldn’t need too 
* MM: Not for sqrt. For functions with asymptote at zero may not may not. Sqrt is defined at 0 to be 0.
* KG: Same for logarithms
* AB: They have an asymptote
* KG: invert sqrt and log and log2 would be infinities
* AB: Why is nan only issue? We can’t represent infinity. They’re just as undefined.
* KG: Don’t think they’re just as undefined. My view is that when you write 1/0 in WGSL something should happen. What should happen is try to gen infinity and if it doesn’t support then we figure out what happens. We should reduce these to what happens when you divide 1/0 or -1/0.
* AB: Why is that different from sqrt(-1) if don’t support nan ti will do op and give result.
* KG: Because infinity is less undefined then nan.
* AB: Don’t know what that means
* MM: Different thought, more concerned about functions which according to IEEE produce nans (or non-normal) for around half or more of domains. View infinity as special case as 1 bit pattern. For sqrt half of domain would produce non-normal number.
* KG: Things making non-normal numbers there isn't’ much you can do with those. Tainted forever. Infinity has more use. Will get use where have revision of WGSL where infinity is real. (haha). Infinities are supported now.
* AB: When you get there I think you’ll get real Nan as well.
* KG: THen what do you want?
* AB: If infinity is undefined then why define nan? Why not just say it’s undefined? If someone mis-uses a math why is that an issue? Do the training wheels go on for nan but not other things? Don’t understand why so special
* MM: Training wheels for functions which are not normal for half or more of input
* AB: Seems arbitrary.
* KG: It is.
* AB: It’s just a judgment call
* KG: We’re in the business of making calls. One might be to weight the evidence and say we don’t or say we do it based on incomplete evidence. Think these things are just not super important to fix. Problems folks already have. Existing code will continue to work. Would it be nice to do better? Probably. If we do better in the future, folks who’s code works continues. We’ll be normalize code which isn’t supposed to work. How strongly do folks want to have to grind out the consensus?
* AB: What described isn't’ quite right. If given behavior, can’t given nan on the future as it maybe incompatible. Not something we can just change the other way. Leaving undefined, restricting to make a single answer is OK.
* MM: In the future when we rely on IEEE 754 and turn off fast math that will be hardware dependent. There is hardware without non-normal numbers. So has to be an extension so text of extension can say changing behavior of ops to return nans as we can no represent them.
* KG: Want to chart course to future where we can ignore nans and infinities that didn’t work. Want a future where they just work when we have support. Would be nice if that behavior could be zero overhead which is to say that absent drivers severely misbehaving we give them ops and they give us results and when we ask for IEEE version, they give IEEE results.
* MM: Think that’s consistent. When the future “make math work” according to IEEE extension teh compiler would not max or round. Fits naturally.
* KG: People then have porting issues at that time?
* MM: That’s point of extension, it’s opt-in and behaviour of program changes.
* KG: Ok with that.
* AB: My opinion, not doing anyone a favour by defining these things.
* MM: Motivation is both JS and webassembly are more defined then what we’d have without the changes we’re discussion.
* KG: They have nan and infinity
* MM: Webassembly has nans but no defined bit pattern so acos(2) can’t rely on the bit pattern but can rely on it being nan and flow naturally from there. JS does define the bit pattern. For us without these changes who knows what you’ll get which is very different then prior art on the web.
* KG: Different from CPU prior art. GPU side we just call the functions and you get what you get. In a way, asking to do better then WebGL.
* MM: Yes, asking to be better then WebGL.
* KG: Don’t think it’s necessary to be better.
* AB: Agree with that, haven’t see lots of complains that folks have this problem with webgl.
* KG: **Think we should table this for now**.

</details>

**Comment by dneto0** [comment]
> MM: Think that’s consistent. When the future “make math work” according to IEEE extension teh compiler would not max or round. Fits naturally.
KG: People then have porting issues at that time?
MM: That’s point of extension, it’s opt-in and behaviour of program changes.

That is a backward incompatibility.  That's broken.  

And it's brokenness-at-a-distance:  It's a globally-scoped option that affects the behaviour of all code in the shader source. That source may be glued together from many different subprojects.

Authors will rely on a restricted behaviour if we give it to them.
If sqrt(x) is really sqrt(max(x,0.0)) then we must assume people will write things like:

      a[i32(sqrt(-15))] = 1;  // and expect this to be the same as a[0]=1

**Comment by dneto0** [comment]
If people write code which tries to do the equivalent of sqrt(-1) or acos(2) then their algorithm is bad and they should fix their algorithm.   The spec should give implementations the room to give them the feedback, when it's possible.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-08-23-minutes#invalid-expressions-in-const-expr-and-override-expr-should-be-error-no-later-than-pipeline-creation-time-3253">WGSL 2022-08-23 Minutes</a></summary>

* KG: Talked about domains of builtins. Feels like we’ve strayed from path. Agree we have to solve these things, worry we’re dealing with what we need to solve and not the level were solving it. Do we agree there should be errors which cause compile errors at const-expr time
* JB: Thumbs up
* DN: Yes
* AB: Don’t know if we all agree
* MM: What is these
* DN: Think we agreed out of bounds index is compile time error. Disagree on domain errors on Floating Point.
* MM: Last week requested splitting this into runtime and compile time and discuss runtime first.
* JB: Agree with MM that sorting out runtime behaviour si fine as let’s settle and make sure other is coherent. What if spec said result is NaN or 0 and that way people don’t depend on it as it maybe nan on supporting platforms. Have upgrade path when GPUs support NaNs and we don’t have unspecified values. All it means is the platforms that can’t support nan have to do some kind of check. I propose this but first object is sqrt is too important and fear for perf reasons we have to accept platform sqrt. Don’t know how to make that call.
* JB: Would fixing alternative to Nan at a value change the way apple feels?
* MM: Would be better then if we did not modify the spec at all. Think that would be acceptable in the spirt of compromise.
* JB: No-one cares about anything but sqrt.
* MM: List of a few of these and it might make sense to split into inverse trig from others. Inverse trig are going to be slow and clamp is fine
* JB: log?
* KG: Want that to be fast
* JB: Really?
* MM: Can probably make resolution about inverse trig functions
* DN: So what JB proposed aligns with what I’m allergic too. Having folks getting used to math working ways it doesn’t. Having that Nan is a good warning. It’s a marker that if you get slow path then you have underspecified GPU and should get a faster one as it’ll be slow
* AB: Doesn't’ seem like market for us as the person suffering has any idea.
* MM: Want to try to understand proposal
    * Premise 1: There is no implementation of vulkan that can handle nans correctly?
        * DN: False
    * Premise 2: Knowing if impl supports nan is extension?
        * DN: True
    * Premise 3: any impl that doesn’t have ext gets slow path, ext is not opt-in would be used. COmpiling making spriv would give Nan when enabled?
* AB: Think DN is wrong about prevalence of vulkan support. Some instructions with core features where you get nan preservation but not applied widely to all instructions
* JB: So just broad absence of support for Nan
* AB: In the spec
* MM: To restate, it is possible for vulkan to say instruction to make nan but not possible to say “and the nan flows into rest of program via IEEE”
* AB: Sort of, more finely grained. Has feature to preserve nans but instructions that feature applies too is not all math instructions.
* JB: So exist instructions where no way to know fi they support nan
* AB: Correct and those are usually GLSL extended math expressions. Can guarantee for core SPIRV instructions like +, *. Other extensions didn't’ get spec, so it’s a hole. Probably not tested any maybe taken advantage of
* JB: So you think they may have skipped nan handling
* AB: If it’s faster, yes.
* KG: Something last week, people said want to do better then WebGL
* MM: WebGL doesn't’ spec out of bounds and I said we should do better then that.
* KG: Need something actionable. What is the thing we’re worried about and scope to prevent. Don’t feel like we have a good handle
* MM: Worried about portability. 
* KG: Means something to you but possibly different to me.
* JB: Program runs same on all browsers? \
KG: Same *. Have big *. How big should * be before v1
* DN: Don’t care about programs misusing math  as being portable or programs misusing indexing
* KG: Think nan-mumbler which takes in nan and spits out elsewhere, Don’t care about preserving that. But care about math being fast.
* MM: So think it’s reasonable to discuss instructions separate from each other
* KG: On webgl side haven’t seen this come up. Believe portability issues are true in theory and we could make test cases but isn’t something we run into in webgl
* JB: Interesting data point. When being absolutist on portability the motivation is what is the experience of folks using. Pragmatic. No principled objection to leaving holes just will it fly and folks be happy. If it’s the case where no-one in webgl has struggled then that is counter to what I was expecting. Weakens my will to struggle about it.
* KG: There are collaborative market forces. Systemic forces are authors want code to run on users machine and we want that too. Make reasonable choices for authors and it works out. Even for portability issue where someone messes up math they didn't’ realize. That’s why I think 1 important question is if we do put our head in the sand, how much can we do that and have it no be dangerous. Can’t tolerate dangerous if we have Undefined Values that cause Undefined Behaviour, can’t have that. But, also the two options are implementation takes on complexity and slowness to make sure undefined values become defined or other option is to do something more structable in spec and forbid them.
* JB: Rewind, what is blocking nan or zero compromise aside from nan not happening on most platforms. Was something I was expecting in future but less confident it will be a thing but could still keep that and still has value in spec as a threat. Sounds like nan or 0 is best option and something folks are ok with
* MM: Yea, could make the spec say this operation will return a nan. If you pass the results into isnan it would return true. What we can’t do is go further then that and say if you flow it through other methods it will be nan
* AB: We removed isnan because we couldn't guarantee them
* MM: Because there are no nans
* JB: Chance the omission in GLSL extension could get filled in?
* AB: Possible to fix in extension but not core spec as a bug fix.
* MM: Instead of emitting those instructs, could emit more instructions
* AB: Don't’ want to emit taylor expansion of our own.
* JB: Those methods are …. Complicated.
* KG: Implementations of complicated functions in terms of complicated functions. Taylor expansion is core common thing
* JB: Do a lot of stuff to make it palatable
* KG: And to make fast. Lot of opcodes in SPIR-V.
* AB: Think KG is persuasive in that this hasn’t been a problem before. Feel this is state of the art for graphics and is kinda expected. Not sure if  we have to work around and define the behaviour and say we’ll give a value instead of nan.
* MM: Maybe mis-understood, when said emit instruction or block of code, didn’t mean emit polyfil, was saying emit clamp and then instruction. Don’t think we should emit our own cosine function.
* KG: If I was the decider of should we fix this before v1, I would vote no. Want to know what we need to fix before v1 release. Agree it would be nice to do better then this to see and experiment and if we could offer on-par performance with better portability guarantees but don’t want to worry about it for v1. Even if could be a cool post-v1 project. What are the v1 requirements.
* JB: Just math or entire issue?
* KG: Entire thing
* JB: Have consensus on indexing.
* KG: Don’t want to unwind consensus. It’s great, but I don't’ really care which way it goes. We should try to wind down our active development. We can make this better, we’d be approximating perfection.
* BC: We could go and wrap the math functions to be more portable but they are the foundation of what most folks will be doing. Impacting perf for portability will likely upset vast majority of folks who don’t care about extremities of functions. Will be some folks who do care. Problematic drivers and hardware should get better. Say we should leave as are and have extension for better math library with full representation for inf and nan. An extension for devices which do support it.
* KG: Few more thoughts then table? ……
* DS: **Can someone split the issue**, we’ve mentioned a few times
* **MM: I’ll do that**.

</details>

**Comment by litherum** [comment]
Split this out into https://github.com/gpuweb/gpuweb/issues/3365 which I believe blocks this issue.

**Comment by dneto0** [comment]
Summarizing for myself (not the official minutes):
- for compile-time evaluation, domain errors (as captured [above](https://github.com/gpuweb/gpuweb/issues/3253#issuecomment-1215906330))  are shader-creation errors.   

Already the spec says, where "one of these types" is AbstractInt and AbstractFloat from the previous paragraph:

> An evaluation of an expression in one of these types [must not](https://gpuweb.github.io/gpuweb/wgsl/#shader-creation-error) overflow or produce infinite, NaN, or undefined results.

Treatment of domain errors at runtime is distinct, that's what #3365 is for, and we agreed it could be different.

So as @jimblandy suggested the spec needs:
- to define a "domain error"
- builtin functions and expressions need to say the conditions inducing a domain error
- say that an const-expression producing a domain error results in a shader creation error

**Comment by dneto0** [comment]
Also, we said:
- divide by zero, in integers is invalid in a const-expr

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-08-30-minutes#invalid-expressions-in-const-expr-and-override-expr-should-be-error-no-later-than-pipeline-creation-time-3253">WGSL 2022-08-30 Minutes</a></summary>

* DS: Which half is this
* MM: Split into runtime vs compile time and we want to runtime behaviour first. This is compile time and we should talk about the other first.

</details>

**Comment by dneto0** [comment]
Compile-time side has all landed, I believe, and is not blocked by runtime side.
Close this one?

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-10-11-minutes#invalid-expressions-in-const-expr-and-override-expr-should-be-error-no-later-than-pipeline-creation-time-3253-">WGSL 2022-10-11 Minutes</a></summary>

* DN: At some time, split into runtime vs compile time. All changes to pin down compile time have landed so think this is candidate for closing. Would like to close.
* KG: Sounds right to me.
* DN: **Cool, closed**.

</details>
